### PR TITLE
Add nullable to Auth.Abstractions, and Auth.Core

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

File: src/Http/Authentication.Abstractions/src/AuthenticationSchemeBuilder.cs [hunk 1]
```diff
@@ -27,17 +27,25 @@ public AuthenticationSchemeBuilder(string name)
...
         /// <summary>
         /// Builds the <see cref="AuthenticationScheme"/> instance.
         /// </summary>
         /// <returns></returns>
-        public AuthenticationScheme Build() => new AuthenticationScheme(Name, DisplayName, HandlerType);
+        public AuthenticationScheme Build()
+        {
+            if (HandlerType is null)
+            {
+                throw new InvalidOperationException($"{nameof(HandlerType)} must be configured to build an {nameof(AuthenticationScheme)}.");
```
Comment (HaoK) [inline src/Http/Authentication.Abstractions/src/AuthenticationSchemeBuilder.cs:45]: I think we lost the type check for IAuthenticationHandler, that was removed from the scheme ctor

```
            if (!typeof(IAuthenticationHandler).IsAssignableFrom(handlerType))
            {	            {
                throw new ArgumentException("handlerType must implement IAuthenticationHandler.");	                throw new ArgumentException("handlerType must implement IAuthenticationHandler.");
            }
```

Comment (pranavkm) [reply src/Http/Authentication.Abstractions/src/AuthenticationSchemeBuilder.cs:45]: That happens as part of the `AuthenticationScheme`: https://github.com/dotnet/aspnetcore/pull/22541/files#diff-5a26792de4ce3d4e74ffdf686c7e9ec1L31.

Comment (HaoK) [reply src/Http/Authentication.Abstractions/src/AuthenticationSchemeBuilder.cs:45]: Right sorry I got confused looking at why this code added here, I thought it got removed, so why do we need to check for null on HandlerType here as well in the ctor

Comment (pranavkm) [reply src/Http/Authentication.Abstractions/src/AuthenticationSchemeBuilder.cs:45]: The ctor no longer allows null-values (it previously checked for nulls) as part of the ctor which continues to stay. Having the annotation makes that more explicit. However the builder now has the choice of using the forgiveness operator or doing something explicit when it encounters a null value. I prefer the latter, but I'll leave it up to you.

Comment (HaoK) [reply src/Http/Authentication.Abstractions/src/AuthenticationSchemeBuilder.cs:45]: I guess my only questions is whether this means we can remove the handlerType == null check from the ctor of AuthenticationScheme?

Comment (pranavkm) [reply src/Http/Authentication.Abstractions/src/AuthenticationSchemeBuilder.cs:45]: That's still constructible by user code. It doesn't really hurt us to have the check in there

**Review by HaoK — APPROVED** [review]

File: src/Http/Authentication.Abstractions/src/IAuthenticationService.cs [hunk 2]
```diff
@@ -18,7 +18,7 @@ public interface IAuthenticationService
         /// <param name="context">The <see cref="HttpContext"/>.</param>
         /// <param name="scheme">The name of the authentication scheme.</param>
         /// <returns>The result.</returns>
-        Task<AuthenticateResult> AuthenticateAsync(HttpContext context, string scheme);
+        Task<AuthenticateResult?> AuthenticateAsync(HttpContext context, string? scheme);
```
Comment (Tratcher) [inline src/Http/Authentication.Abstractions/src/IAuthenticationService.cs:21]: Where would AuthenticateAsync return a null value?

Comment (pranavkm) [reply src/Http/Authentication.Abstractions/src/IAuthenticationService.cs:21]: https://github.com/dotnet/aspnetcore/blob/master/src/Http/Authentication.Core/src/AuthenticationService.cs#L81-L98

Comment (Tratcher) [reply src/Http/Authentication.Abstractions/src/IAuthenticationService.cs:21]: That seems precautionary, it doesn't look like any of the handlers return null from the call to AuthenticateAsync on line 80.

Comment (pranavkm) [reply src/Http/Authentication.Abstractions/src/IAuthenticationService.cs:21]: I was walking back the annotation based on the implementation. We could make it non-null if we expect this to be non-null.

Comment (Tratcher) [reply src/Http/Authentication.Abstractions/src/IAuthenticationService.cs:21]: Yeah, try it. It looks like IAuthenticationHandler.AuthenticateAsync is already non-nullable.

**Pull request merged**